scenario_player/releases.py

```diff
@@ -0,0 +1,336 @@
+import pathlib
+import shutil
+import stat
+import sys
+from tarfile import TarFile
+from typing import Dict, List, Tuple, Union
+from zipfile import ZipFile
+
+import requests
+import structlog
+from cachetools.func import ttl_cache
+
+log = structlog.get_logger(__name__)
+
+
+RAIDEN_RELEASES_URL = 'https://raiden-nightlies.ams3.digitaloceanspaces.com/'
+
+
+class PLATFORM_SEPCIFIC_VARS:
+    @classmethod
+    def latest_file_name(cls) -> str:
+        """Construct the file name of the `latest` file on our cloud."""
+        if sys.platform == "darwin":
+            return '_LATEST-macOS-x86_64.txt'
+        return '_LATEST-linux-x86_64.txt'
+
+    @classmethod
+    def archive_name_template(cls) -> str:
+        """Construct the archive name for the release, factoring in the system's platform."""
+        if sys.platform == "darwin":
+            return 'raiden-v{version}-macOS-x86_64.zip'
+        return 'raiden-v{version}-linux-x86_64.tar.gz'
+
+
+@ttl_cache(maxsize=1, ttl=600)
+def get_latest_release():
+    """Retrieve the latest release's URL path"""
+    url = RAIDEN_RELEASES_URL + PLATFORM_SEPCIFIC_VARS.latest_file_name()
+    log.debug('Fetching latest Raiden release')
+    response = requests.get(url)
+    text = response.text
+    return text.strip()
+
+
+def is_executable(path: pathlib.Path) -> bool:
+    """Check for a set executable bit on the given pathlib.Path object."""
+    return path.stat().st_mode & stat.S_IXUSR == stat.S_IXUSR
+
+
+class ReleaseArchive:
+    """Wrapper class for extracting a Raiden release from its archive.
+
+    Supplies a context manager and file-type detection, which allows choosing
+    the correct library for opening the archive automatically.
+    """
+
+    def __init__(self, path: pathlib.Path) -> None:
+        self.path = path
+        self._context = None
+        self.validate()
```

Comment:
`validate` works on `self.files`, which requires `self._context`. But `_context` is still `None` at this point.